Analysis/LeptonID/src/LeptonIDProcessor.cc

```diff
@@ -0,0 +1,374 @@
+
+    registerProcessorParameter("UsedVariables",
+                               "List of variables used during training",
+                               _usedVariables,
+                               usedVars);
+}
+
+void LeptonIDProcessor::setupTree()
+{
+    _tree->Branch("truePDG", &_truePDG);
+    _tree->Branch("parentPDG", &_parentPDG);
+
+    for (auto &v : _vars)
+    {
+        _tree->Branch(v.first.c_str(), &(v.second));
+    }
+}
+
+void LeptonIDProcessor::setupMVAReader()
+{
+
+    for (const auto &varname : _usedVariables) {
+        _mvareader.AddVariable(varname, &_vars[varname]);
+    }
+    // XXX: the reader also needs the spectator variables but does not use them...
+    _mvareader.AddSpectator("abs(truePDG)", (int*)nullptr);
+    _mvareader.AddSpectator("abs(parentPDG)", (int*)nullptr);
+    _mvareader.AddSpectator("seenP", (float*)nullptr);
+    _mvareader.AddSpectator("seenCosTheta", (float*)nullptr);
+
+    _mvareader.BookMVA(_mvaname, _weightfile);
+}
+
+void LeptonIDProcessor::init()
+{
+    streamlog_out(DEBUG) << "   init called  " << std::endl;
+    if (_buildTree) {
+        // if I keep this here it magically appears in the aida file :)
+        _tree = new TTree("pfos", "pfos");
+        setupTree();
+    }
+
+    if (_evalMVA) {
+        setupMVAReader();
+    }
+
+}
+
+void LeptonIDProcessor::clear()
+{
+    _truePDG = 0;
+    _parentPDG = 0;
+
+    for (auto v : _vars) {
+        v.second = 0.0f;
+    }
+}
+
+void LeptonIDProcessor::processRunHeader(LCRunHeader *run)
+{
+    (void) run;
+}
+
+void LeptonIDProcessor::modifyEvent(LCEvent *evt)
+{
+    streamlog_out(DEBUG) << "process Event called" << std::endl;
+    LCCollection *PandoraPFOs = nullptr;
+    LCCollection *InputPFOs = nullptr;
+    LCCollection *RecoMCTruthLink = nullptr;
+    LCCollection *MCTruthRecoLink = nullptr;
+    LCCollection *PandoraClusters = nullptr;
+
+    try
+    {
+        PandoraPFOs = evt->getCollection(_PandoraPFOsColName);
+        InputPFOs = evt->getCollection(_InputPFOsColName);
+        PandoraClusters = evt->getCollection(_PandoraClustersColName);
+        if (_buildTree) {
+            RecoMCTruthLink = evt->getCollection(_RecoMCTruthLinkName);
+            MCTruthRecoLink = evt->getCollection(_MCTruthRecoLinkName);
+        }
+    }
+    catch (lcio::DataNotAvailableException const& e)
+    {
+        streamlog_out(WARNING) << "atleast one collection not available" << std::endl;
+        streamlog_out(WARNING) << e.what() << std::endl;
+        return;
+    }
+
+    //LCRelationNavigator RecoMCTruthNavigator(RecoMCTruthLink);
+    //LCRelationNavigator MCTruthRecoNavigator(MCTruthRecoLink);
+    std::unique_ptr<LCRelationNavigator> RecoMCTruthNavigator;
+    std::unique_ptr<LCRelationNavigator> MCTruthRecoNavigator;
+    if (_buildTree) {
+        RecoMCTruthNavigator = std::make_unique<LCRelationNavigator>(RecoMCTruthLink);
+        MCTruthRecoNavigator = std::make_unique<LCRelationNavigator>(MCTruthRecoLink);
+    }
```

Comment:
Not sure I immediately see why these cannot live on the stack. What am I missing?

Reply:
Here my intention was to only initialize the `LCRelationNavigator` only if necessary to not depend on the existence of `MCParticle` and relation information when not needed. I first tried something like

```c++
LCRelationNavigator RecoMCTruthNavigator;
LCRelationNavigator MCTruthRecoNavigator;
  if (_buildTree) {
    RecoMCTruthNavigator = RecoMCTruthNavigator(RecoMCTruthLink);
    MCTruthRecoNavigator = MCTruthRecoNavigator(MCTruthRecoLink);
  }
```
but that doesn't work. This did but I am open to suggestions.. :)

Reply:
Ah right, protected default constructor will do that.